pdf/pdflinkarea.cpp

```diff
@@ -148,8 +196,53 @@ void PDFLinkArea::mouseReleaseEvent(QMouseEvent *event)
     if (!rect.contains(event->pos()))
         return;
 
+    // Click action logic in order.
+    // - click on selection;
+    // - unselect if selection is set;
+    // - click on annotation;
+    // - click on link;
+    // - click.
+
     if (!d->canvas) {
         emit clicked(d->clickLocation);
+    } else if (d->clickOnSelection) {
+        emit selectionClicked();
+    } else if (d->selection && d->selection->count() > 0) {
+        d->selection->unselect();
+    } else if (d->annotation != nullptr) {
+        // proxy will be child of this object to avoid memory leak.
+        // Todo: transfer ownership to QML, if possible. Before that,
+        //       all created proxy objects will be alive up to the moment
+        //       the document object is released.
+        switch (d->annotation->subType()) {
+        case Poppler::Annotation::SubType::AText: {
+            PDFTextAnnotation *proxy = new PDFTextAnnotation
+                (static_cast<Poppler::TextAnnotation*>(d->annotation),
+                 d->canvas->document(), d->pressedBox.first, this);
+            emit annotationClicked(proxy);
+            return;
+        }
+        case Poppler::Annotation::SubType::ACaret: {
+            PDFCaretAnnotation *proxy = new PDFCaretAnnotation
+                (static_cast<Poppler::CaretAnnotation*>(d->annotation),
+                 d->canvas->document(), d->pressedBox.first, this);
+            emit annotationClicked(proxy);
+            return;
+        }
+        case Poppler::Annotation::SubType::AHighlight: {
+            PDFHighlightAnnotation *proxy = new PDFHighlightAnnotation
+                (static_cast<Poppler::HighlightAnnotation*>(d->annotation),
+                 d->canvas->document(), d->pressedBox.first, this);
+            emit annotationClicked(proxy);
+            return;
+        }
+        default: {
+            PDFAnnotation *proxy = new PDFAnnotation
+                (d->annotation, d->canvas->document(), d->pressedBox.first, this);
+            emit annotationClicked(proxy);
+            return;
```

Comment:
Just a minor thing, but all cases emit same signal, can as well do that outside the switch.